### PR TITLE
Support Keycloak Dev Service when OIDC client is used without OIDC extension

### LINKED ISSUE #43086: Make Keycloak dev service independent of OIDC extension
### Description

Scenario:
A Microservice receives requests that don't use OIDC (unauthenticated or authenticated through a different/external mechanism), then creates requests through a client towards an OIDC-protected service.

In this case, if the user only adds **quarkus-oidc-client** and **quarkus-rest-client-oidc-filter**, the keycloak dev service is not enable. On the other hand, if they add **quarkus-oidc** as well, then OIDC authentication is enabled for all resources by default.

To work around this, the user can use `quarkus.http.auth.proactive=false`, as mentioned in [this discussion](https://github.com/quarkusio/quarkus/discussions/43075). This, however is not clear from the documentation and it is an extra step, which makes it counter-intuitive. The user also has to switch the oidc extension off for the final jar and even doing so, the extension is still going to be bundled with the executable.

Enabling the dev service without requiring the **quarkus-oidc** package will in my opinion improve the development experience, so please consider this request.

### Implementation ideas

The Dev service could be made into a separate package and included by both **quarkus-oidc** and **quarkus-oidc-client** in case they are used separately.

### PR DESCRIPTION
**Pull request by michalvavrik**
- Closes: https://github.com/quarkusio/quarkus/issues/43086
- OIDC Client will be started and preconfigured without the OIDC extension being present
- removing `OidcDevServicesBuildItem`; the class wasn't final, therefore could never be used as SimpleBuildItem must be final https://github.com/quarkusio/quarkus/blob/4cab5df30997c424b269858ce30d186d74c61da7/core/builder/src/main/java/io/quarkus/builder/item/BuildItem.java#L22; extensions can set `quarkus.keycloak.devservices.enabled` with SR Config during the build time
- `grant` was deprecated on Dev Svc config from 2021 and since I am moving this config, I am also dropping this property
- OIDC Client tab is only added and contains Keycloak Admin URL:
image

### CODE REVIEW

**Comment by michalvavrik** [comment]
cc @sberyozkin but let's see how CI goes, I run couple of extension unit tests locally, but not every integration test

**Comment by michalvavrik** [comment]
`OidcClientTooManyJwtCredentialKeyPropsTestCase` became flaky and I cannot say if it is consequence of this PR, but I'll have a look into it. The `integration-tests/oidc-client-wiremock ` I didn't run yesterday so I'll look into them.

**Comment by michalvavrik** [comment]
So `OidcClientTooManyJwtCredentialKeyPropsTestCase ` was about missing `quarkus.oidc.auth-server-url` being thrown sooner than other expected validation exception. And `integration-tests/oidc-client-wiremock ` should not start Dev Service because it is a wiremock. I'll push fixes.

**Comment by michalvavrik** [comment]
I'll wait for https://github.com/quarkusio/quarkus/pull/43601 merge before I push fixes because this needs to adapt changes there anyway (I am moving file that is edited there).

**Comment by sberyozkin** [comment]
Thanks very much @michalvavrik, 
I think having an OIDC client dedicate card is worth it, even if OIDC card is present, it is just another link to the Keycloak provider for now which is not really conflicting and then users will start asking for some more OIDC client specific to add there, so will be adding some more....

What happens when both `quarkus-oidc` and `quarkus-oidc-client` are loaded ?

**Comment by michalvavrik** [comment]
> What happens when both quarkus-oidc and quarkus-oidc-client are loaded ?

DEV UI wise - card is not produced in the OIDC Client, so you have OIDC card that also contains that link
DEV Svc wise - sadly it is almost everywhere in our tests (only module that I know of that contains OIDC client and not OIDC is the OIDC CLient wiremock) so you can see it -> nothing bad, dev svc is started just right as previously it did

**Comment by sberyozkin** [comment]
@michalvavrik 

>DEV UI wise - card is not produced in the OIDC Client, so you have OIDC card that also contains that link

But we don't have to start hiding it, I think it is shown now on `main` anyway when OIDC is loaded, with the text only, so this PR only has a KC link added to it.

>dev svc is started just right as previously it did

Sure, I got confused for a moment  if it would start twice, but no, it will be a single instance only

**Comment by michalvavrik** [comment]
> But we don't have to start hiding it, I think it is shown now on main anyway when OIDC is loaded, with the text only, so this PR only has a KC link added to it.

I am not sure we understand each other, let me try again:

- OIDC CLlient has own DEV UI card with KC Link
- OIDC has own DEV UI card with KC link and link to the OIDC DEV UI component

I hide OIDC CLient card when OIDC Is present. If you are saying "let's show it" I think it's alright and I'll do it (please confirm!). I somehow misunderstood your email that said _I guess though when no `quarkus-oidc` is involved, there should be maybe Keycloak card only offering a Keycloak admin link ? It can be done independently though._ I suppose I overthink it.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java [hunk 1]
```diff
@@ -17,7 +17,8 @@ public class OidcClientTooManyJwtCredentialKeyPropsTestCase {
     static final QuarkusUnitTest test = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
                     .addAsResource(new StringAsset(
-                            "quarkus.oidc-client.token-path=http://localhost:8180/oidc/tokens\n"
+                            "quarkus.keycloak.devservices.enabled=false\n"
+                                    + "quarkus.oidc-client.token-path=http://localhost:8180/oidc/tokens\n"
```
Comment (sberyozkin) [inline extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java:21]: Interesting case. This combination is an optimization for any provider which does not support auto-discovery, Instead of typing 3 properties like

```
auth-server-url=some-base-url
discovery-enabled=false
token-path=some-relative-path
```

users just type an absolute URL. I think we need to check for this property as well

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java [hunk 2]
```diff
@@ -83,13 +83,16 @@ public class KeycloakDevServicesProcessor {
     private static final String TENANT_ENABLED_CONFIG_KEY = CONFIG_PREFIX + "tenant-enabled";
     private static final String AUTH_SERVER_URL_CONFIG_KEY = CONFIG_PREFIX + "auth-server-url";
     private static final String PROVIDER_CONFIG_KEY = CONFIG_PREFIX + "provider";
-    // avoid the Quarkus prefix in order to prevent warnings when the application starts in container integration tests
-    private static final String CLIENT_AUTH_SERVER_URL_CONFIG_KEY = "client." + CONFIG_PREFIX + "auth-server-url";
```
Comment (sberyozkin) [inline extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: This one was introduced by @geoand to support the shared network cases, the actual integration test was removed but it would have to be restored, but users may be depending on it

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java:21]: Please, which number do you mean:

1. we check for `quarkus.oidc-client.token-path` and if it is set, we don't start Keycloak Dev Svc? (I think we started it before when OIDC was present)
2. we check fo `quarkus.oidc-client.token-path` and if it is set, we don't set `quarkus.oidc-client.auth-server-url` to just started Keycloak Dev Svc)?

Or maybe something else @sberyozkin ?

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: Yes, I checked that PR and there are no details I could make sense of. So I rewrote logic, I only set OIDC client properties when OIDC client is present and enabled; I only set OIDC properties when OIDC is present and enabled

I cannot imagine it could actually cause any problem and if so, let's have reproducer. I don't really believe that until I'll know more. Sorry.

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java [hunk 3]
```diff
@@ -1,4 +1,7 @@
-package io.quarkus.oidc.deployment.devservices.keycloak;
+package io.quarkus.oidc.common.deployment.devservices.keycloak;
```
Comment (sberyozkin) [inline extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:1]: I was thinking it would actually end up in the devservices set of extensions, here, https://github.com/quarkusio/quarkus/tree/main/extensions/devservices (so far it hosts database specific services only but we can add `oidc` there, may be in the `security` subfolder), so that `keycloak-admin-client` can eventually pick it up too.

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: @geoand this is the PR in discussion https://github.com/quarkusio/quarkus/pull/21999

Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: My personal guess is that `CLIENT_AUTH_SERVER_URL_CONFIG_KEY` was about config validation messages and therefore my changes avoids that.

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:1]: Well, I don't mind. I am careful when creating new modules, I don't create it until there is requirement, but if that's what you want, I'll just change it.

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 2 above)
Comment (geoand) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: I can't say I remember the details, sorry

Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: > I can't say I remember the details, sorry

I'd not expect you to. No worry, I think this comment was misread.

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 3 above)
Comment (sberyozkin) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:1]: Thanks @michalvavrik, IMHO it would be the next requirement from the Keycloak-admin-client users anyway :-), like, if you can do it with OIDC client, why can't we do it with keycloak-admin-client :-). I'm pretty sure we had some related queries

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: I apologize for previous comments @sberyozkin . I invested time into that and misjudged that property. There is no need to do anything before this PR goes in as I'll restore previous behavior. I understand purpose of this property now.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java:21]: Once again I think I misunderstood you (mainly because I don't understand relation to this test you are commenting on), basically you say:

_preset following properties_:

```
auth-server-url=some-base-url
discovery-enabled=false
token-path=some-relative-path
```

but what if discovery should be enabled and auth server URL set? I think current behavior is not breaking. Unless there is something wrong, let's make optimizations later because I'll be frank - I don't understand these properties as you do.

Comment (sberyozkin) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java:21]: Hi @michalvavrik 

>we check for quarkus.oidc-client.token-path and if it is set, we don't start Keycloak Dev Svc? 

yes, unless OIDC is available

> (I think we started it before when OIDC was present)

`quarkus-oidc` does not support omitting `auth-server-url` right now, as it has at least 3 endpoints configured otherwise, this optimization is only for oidc-client (oidc-client-registration)

> we check fo quarkus.oidc-client.token-path and if it is set, we don't set quarkus.oidc-client.auth-server-url to just started Keycloak Dev Svc)?

It should not be started

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 2 above)
Comment (sberyozkin) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:87]: Hey @michalvavrik, there is absolutely no need to apologize :-), your help/time is super appreciated, we just discuss and challenge some comments as usual in a normal way, thanks

File: extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/oidc-common/deployment/src/main/java/io/quarkus/oidc/common/deployment/devservices/keycloak/KeycloakDevServicesProcessor.java:1]: I have separated it (still testing locally) I just want to clarify I won't provide Dev Svc for KC Admin client ATM, it will give us opportunity to implement it with one trivial PR, but we need to  keep this scope because it has potential for merge conflicts or on missing something that will change between now and merging.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java]: should be done; it seems to address the flakiness, but I am still on fence so I improved error message in these 2 tests to tell us more if this ever happens

Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java]: `discovery-enabled=false` fails too many tests, I am tired of adding `discovery-enabled=true` as I realize you would probably not like it anyway; this is PR that can conflict with others. I'd suggest to make optimizations like this in a separate PR that will be easier to rebase etc. Thank you

Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientTooManyJwtCredentialKeyPropsTestCase.java]: Of course I'll fix any proposal and fixes you will suggest, I just suggest to leave enhancements and optimizations separately.

File: extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java [hunk 4]
```diff
@@ -184,6 +191,27 @@ private AccessTokenInstanceBuildItem build() {
         return index.getIndex().getAnnotations(ACCESS_TOKEN).stream().map(ItemBuilder::new).map(ItemBuilder::build).toList();
     }
 
+    @BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class)
```
Comment (sberyozkin) [inline extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java:194]: Proposing to move it to a dedicated `OidcClientDevServiceBuildStep`, may be we can later share it with KeycloakAdminClient, etc

File: extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java [hunk 5]
```diff
@@ -184,6 +191,27 @@ private AccessTokenInstanceBuildItem build() {
...
+    KeycloakDevServicesRequiredBuildItem requireKeycloakDevService() {
+        // this needs to be done as OIDC Common doesn't know if the OIDC Client is enabled
```
Comment (sberyozkin) [inline extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java:196]:
```suggestion       
```

File: extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java (hunk 4 above)
Comment (sberyozkin) [reply extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java:194]: @michalvavrik never mind, let's keep it here for now, I can check later if sharing some of it makes sense

File: extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java (hunk 5 above)
Comment (sberyozkin) [inline extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java:196]:
```suggestion
        // this needs to be done as the shared Keycloak devservice doesn't know if the OIDC Client is enabled
```

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/OidcBuildStep.java [hunk 6]
```diff
@@ -395,6 +398,12 @@ List<HttpAuthMechanismAnnotationBuildItem> registerHttpAuthMechanismAnnotation()
                 new HttpAuthMechanismAnnotationBuildItem(DotName.createSimple(BearerTokenAuthentication.class), BEARER_SCHEME));
     }
 
+    @BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class)
+    KeycloakDevServicesRequiredBuildItem requireKeycloakDevService() {
+        // this needs to be done as OIDC Common doesn't know if the OIDC is enabled
```
Comment (sberyozkin) [inline extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/OidcBuildStep.java:403]:
```suggestion
        // this needs to be done as the shared Keycloak devservice doesn't know if the OIDC is enabled
```

File: test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/client/KeycloakTestClient.java [hunk 7]
```diff
@@ -24,10 +24,17 @@
 public class KeycloakTestClient implements DevServicesContext.ContextAware {
 
     private final static String CLIENT_AUTH_SERVER_URL_PROP = "client.quarkus.oidc.auth-server-url";
+
```
Comment (sberyozkin) [inline test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/client/KeycloakTestClient.java:27]: I think `KeycloakTestClient` should remain unchanged,  it is about helping tests to acquire tokens from OIDC server, the OIDC client does not need help from `KeycloakTestClient`. 
Well, may be only `KeycloakTestClient` might help to upload realms, and it is only  `OIDC_CLIENT_AUTH_SERVER_URL_CONFIG_KEY` that would be relevant.

But as far as the non admin token acquisition goes, I'm worried we may get some strange errors when both OIDC server and OIDC client are loaded, `quarkus-oidc` tests for acquiring tokens from OIDC server will pick oidc-client properties pointing to different realms, etc.

As far as this PR is concerned I propose not to deal with KeycloakTestClient updates. We can consider something later

File: extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java [hunk 8]
```diff
@@ -188,104 +178,36 @@ public class DevServicesConfig {
...
      * Memory limit for Keycloak container
      * </p>
      * If not specified, 750MiB is the default memory limit.
      */
-    @ConfigItem(defaultValue = "750M")
-    public MemorySize containerMemoryLimit;
+    @WithDefault("750M")
+    MemorySize containerMemoryLimit();
 
-    @Override
```
Comment (sberyozkin) [inline extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java:266]: Just commenting that that this was done to check if Keycloak devservice may have to be reloaded, I'm yet to check the moved KC devservice code, so not sure yet these checks have moved there

File: test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/client/KeycloakTestClient.java (hunk 7 above)
Comment (michalvavrik) [reply test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/client/KeycloakTestClient.java:27]: It is not a problem, changes in the `KeycloakTestClient` are not required and I'll remove them.

If you don't mind, I think you very much misunderstood changes here, I am simply adding support to use `KeycloakTestClient` when OIDC extension is not present. I understand it won't be frequent as this is OIDC client, but I still can imagine users will want to get token in the test directly.

Anyway, it was just an idea, I'll remove it. Thanks for nice explanation.

File: extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java (hunk 8 above)
Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java:266]: Sure thing, look here https://smallrye.io/smallrye-config/Main/config/mappings/#tostring-equals-hashcode.

Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java:266]: Just for context - I simply removed it as SR Config generates it for us.

File: test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/client/KeycloakTestClient.java (hunk 7 above)
Comment (sberyozkin) [reply test-framework/keycloak-server/src/main/java/io/quarkus/test/keycloak/client/KeycloakTestClient.java:27]: Hi @michalvavrik, how one would use it though if the Quarkus endpoint is not expected to be protected by quarkus-oidc, that is the question. But yeah, we can iterate on it earlier once users tell us a bit more about their OIDC client plus KC devservice experience

File: extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java (hunk 8 above)
Comment (sberyozkin) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesConfig.java:266]: @michalvavrik Cool

**Comment by sberyozkin** [comment]
@michalvavrik It looks quite perfect, minor suggestions are there only, I'll run your PR a bit later as well and then we can merge. Later, I'll follow the same pattern for OIDC dynamic client registration as well and eventually we can cover keycloak-admin-clients

**Review by sberyozkin — APPROVED** [review]
I've tested Michal's PR on my laptop, with `quarkus-quickstarts/security-openid-connect-quickstart` to confirm a KC DevService is started in dev and test modes, and that OIDC DevUI works as usual.
Additionally, I've confirmed OIDc Dev UI offers Auth0 login experience with `quarkus.oidc.auth-server-url` pointing to an Auth0 domain

**Comment by sberyozkin** [comment]
@michalvavrik I thought for a second if it would make sense to keep the base devservice abstract and have extension specific extensions which would deal with their own properties (oidc, oidc-client), etc and have some build time coordination informing each other the the dev service is started.
For example, we also have `keycloak-authorization`, which is OIDC plus KC authorization policy.
But then I thought we can rework easily if necessary

**Comment by michalvavrik** [comment]
> @michalvavrik I thought for a second if it would make sense to keep the base devservice abstract and have extension specific extensions which would deal with their own properties (oidc, oidc-client), etc and have some build time coordination informing each other the the dev service is started. For example, we also have `keycloak-authorization`, which is OIDC plus KC authorization policy. But then I thought we can rework easily if necessary

I am all for loose coupling, did it this way because it feels easier to understand than new level of abstraction. FWIW if it will be me who introduce this for other extensions, I'll move these config properties to the extensions and enhance `keycloakdevservicerequired` build item. I can draw it if you have issue to migrate some service, for example `keycloak-authorization`.

**Comment by sberyozkin** [comment]
@michalvavrik Sure, we can look at it all later, thanks

**Comment by rsvoboda** [comment]
Almost feels like there should be a not in documentation about this @michalvavrik .

I also checked https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.16 and https://quarkus.io/blog/quarkus-3-16-1-released/. I didn't find anything highlighting this change.

**Comment by michalvavrik** [comment]
> Almost feels like there should be a not in documentation about this @michalvavrik .

If you mean https://quarkus.io/version/main/guides/security-openid-connect-dev-services then I agree. This week I'll be implementing Keycloak Dev Svc for Keycloak Admin Clients, so I'll write into that document that we now configure auth-server-url for default OIDC client and client registration (and that we configure KC with anonymous client registration policy).

> I also checked https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.16 and https://quarkus.io/blog/quarkus-3-16-1-released/. I didn't find anything highlighting this change.

"Breaking" part is that container is started if you use OIDC client without OIDC extension and you don't have configured auth server url for default client and you don't have disabled dev svc (or kc dev svc). I can see it can affect you if you don't have docker. But actual users - if they have OIDC Client enabled, they probably need auth server, so this can have only very little impact.

I'll let @sberyozkin deal with migration guide as I don't have perms.
